substantial/backends/redis.py

```diff
@@ -11,70 +9,133 @@
 class RedisBackend(Backend):
     def __init__(self, host: str, port: int, **kwargs):
         self.redis = redis.Redis(host=host, port=port, decode_responses=True, **kwargs)
+        self.separator = ":/"
+        self.base_prefix = "substantial"
         # self.redis.flushall()
 
+    # Utils
+    def _key(self, *parts: str) -> str:
+        invalid_chunks = [part for part in parts if self.separator in part]
+        if len(invalid_chunks) > 0:
+            raise ValueError(
+                f"Fatal: parts {invalid_chunks} cannot contain separator '{self.separator}'"
+            )
+        return f"{self.base_prefix}{self.separator}{self.separator.join(parts)}"
+
+    def _parts(self, key: str) -> List[str]:
+        if not key.startswith(self.base_prefix):
+            raise ValueError(
+                f"Invalid key '{key}': required prefix '{self.base_prefix}' not present"
+            )
+
+        return key[len(self.base_prefix) :].split(self.separator)
+
+    # Backend
+
     async def read_events(self, run_id: str) -> Union[Records, None]:
-        key = "_".join(["runs", run_id, "events"])
-        val = self.redis.get(key)
+        event_key = self._key("runs", run_id, "events")
+        val: str | None = self.redis.register_script("""
+            local event_key = KEYS[1]
+            if redis.call("EXISTS", event_key) == 1 then
+                return redis.call("GET", event_key)
+            else
+                return nil
+            end
+        """)(keys=[event_key])
+
         return None if val is None else Records().from_json(val)
 
     async def write_events(self, run_id: str, content: Records) -> None:
-        key = "_".join(["runs", run_id, "events"])
+        key = self._key("runs", run_id, "events")
         self.redis.set(key, content.to_json(indent=4))
 
     async def read_all_metadata(self, run_id: str) -> List[str]:
-        base_key = "_".join(["runs", run_id, "logs"])  # queue
-        ret = []
-        for schedule in self.redis.lrange(base_key, 0, -1):
-            log = self.redis.get("_".join[run_id, schedule])
-            # could be polymorphic
-            ret.append(log)
-        return ret
+        log_key = self._key("runs", run_id, "logs")
+        logs = self.redis.register_script("""
+            local log_key = KEYS[1]
+            local run_id = ARGV[1]
+
+            local sched_keys = redis.call("LRANGE", log_key, 0, -1)
+            local logs = {}
+            for _, sched_key in ipairs(sched_keys) do
```

Comment:
when you use Lua you make concurrent request wait, I wonder wether a lrange + mget would not be a better fit here. Similarly for the follow ups once, I would rely on Lua only if there is no other command required

Reply:
> when you use Lua you make concurrent request wait
This happens on a specific run but not on all of them so I was not really worried on the performance. The other complex lua scripts follow similar rationals.

> I would rely on Lua only if there is no other command required
I think it still matters if multiple wites are involved or if in between two value retrieval there is no delete from outside.

> lrange + mget
That could actually work! Also this is all read ops without potential deletes in between queries so we can get rid of Lua in this situation I guess.